src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1354,13 +1357,156 @@ public ValidateCountAttribute(int minLength, int maxLength)
     }
 
     /// <summary>
+    /// Base class for all valid values generators with cache support.
+    /// </summary>
+    public abstract class CachedValidValuesGeneratorBase : IValidateSetValuesGenerator
+    {
+        // Cached valid values.
+        private string[] _validValues;
+
+        /// <summary>
+        /// Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.
+        /// By default 'ValidValuesCacheExpiration = 0' and no valid values is cached (is re-evaluated every time it is used).
+        /// </summary>
+        public int ValidValuesCacheExpiration { get; set; } = 0;
+
+        /// <summary>
+        /// Abstract method to generate a valid values.
+        /// </summary>
+        public abstract string[] GenerateValidValues();
+
+        /// <summary>
+        /// Get a valid values.
+        /// </summary>
+        public IEnumerable<string> GetValidValues()
+        {
+            if (_validValues != null)
+            {
+                return _validValues;
+            }
+
+            var _validValuesNoCache = GenerateValidValues();
+
+            if (_validValuesNoCache == null)
+            {
+                throw new ValidationMetadataException(
+                    "ValidateSetGeneratedValidValuesListIsEmpty",
+                    null,
+                    Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+            }
+
+            if (ValidValuesCacheExpiration > 0)
+            {
+                Task.Delay(ValidValuesCacheExpiration).ContinueWith((task) => _validValues = null);
+                _validValues = _validValuesNoCache;
+            }
+
+            return _validValuesNoCache;
+        }
+    }
+
+    /// <summary>
     /// Validates that each parameter argument is present in a specified set
     /// </summary>
     [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
     public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
     {
+        // We can use either static '_validValues'
+        // or dynamic valid values list generated by instance of 'validValuesGeneratorType'.
         private string[] _validValues;
 
+        // The valid values generator cache works across 'ValidateSetAttribute' instances.
+        private static ConcurrentDictionary<Type, ValidValuesGeneratorCacheEntry> s_ValidValuesGeneratorCache = new ConcurrentDictionary<Type, ValidValuesGeneratorCacheEntry>();
```

Comment:
Restarting PowerShell is an occasional annoyance, but I'm not sure this will make a difference.

For C# implementations, you need to restart anyway.
For script implementations - I think there are acceptable ways to develop your generator and test it without restarting PowerShell.

So my preference is to keep the code as simple as possible and only introduce additional complexity after gaining more experience with the feature and determining it really is necessary.